src/libraries/System.Net.Requests/src/System/Net/HttpWebRequest.cs

```diff
@@ -1693,11 +1693,16 @@ private static HttpClient CreateHttpClient(HttpClientParameters parameters, Http
                 {
                     // This is legacy feature and we don't have public API at the moment.
                     // So we want to process it only if explicitly set.
-                    var settings = typeof(SocketsHttpHandler).GetField("_settings", BindingFlags.NonPublic | BindingFlags.Instance)?.GetValue(handler);
-                    Debug.Assert(settings != null);
-                    FieldInfo? fi = Type.GetType("System.Net.Http.HttpConnectionSettings, System.Net.Http")?.GetField("_impersonationLevel", BindingFlags.NonPublic | BindingFlags.Instance);
-                    Debug.Assert(fi != null);
-                    fi.SetValue(settings, request.ImpersonationLevel);
+                    GetImpersonationLevel(GetSettings(handler)) = request.ImpersonationLevel;
+
+                    const string HttpConnectionSettingsTypeName = "System.Net.Http.HttpConnectionSettings, System.Net.Http";
+
+                    [UnsafeAccessor(UnsafeAccessorKind.Field, Name = "_settings")]
+                    [return: UnsafeAccessorType(HttpConnectionSettingsTypeName)]
+                    static extern ref object GetSettings(SocketsHttpHandler handler);
```

Comment:
Is the `_settings` field typed as `object`? Remember that byrefs don't support variance so this is likely to fail unless `_settings` is an `object`.

Reply:
> Is the `_settings` field typed as `object`? Remember that byrefs don't support variance so this is likely to fail unless `_settings` is an `object`.

Would it make sense to allow `readonly ref` with variance in UnsafeAccessor then and cast away the readonlyness with unsafe if needed?

Reply:
We have plans instead to eventually use `TypedReference` for this scenario.